### PR TITLE
upstream: allow extension protocol options to be used with http filters

### PR DESCRIPTION
**Pull request by yuval-k**
https://github.com/envoyproxy/envoy/pull/4098 Added protocol extensions for clusters and network filters. This PR expands this to http filters as well (this can be used for example in a filter that translates http to a different protocol, like our NATS Streaming filter)

Risk Level: low, ignored by existing protocols
Testing: unit tests
Docs Changes: inline
Release Notes: n/a

Few notes:
* I've extracted createProtocolOptionsConfig, createEmptyProtocolOptionsProto to an interface called ProtocolOptionsConsumer as the factories that inherit from it are the ones that will eventually consume the configuration (via created filters). not sure if its the best name, suggestions are welcome.
* when building the config map, for each filter name, I first check if there's a network filter and if there isn't, i check if there's an http filter. I did not handle the case of both existing as this should never happen (as filter names are namespaced by convention).
* Updated existing unit tests to test both http and network filters.

### CODE REVIEW

**Comment by mattklein123** [comment]
Make sense to me. @zuercher PTAL

File: include/envoy/server/filter_config.h [hunk 1]
```diff
@@ -311,6 +323,16 @@ class NamedHttpFilterConfigFactory {
     return nullptr;
   }
 
+  // ProtocolOptionsConsumer
+  Upstream::ProtocolOptionsConfigConstSharedPtr
+  createProtocolOptionsConfig(const Protobuf::Message& config) override {
```
Comment (zuercher) [inline include/envoy/server/filter_config.h]: What do you think about moving the default implementations (here and from the NamedNetworkFilterConfigFactory) to ProtocolOptionsConsumer?

File: test/common/upstream/upstream_impl_test.cc [hunk 2]
```diff
@@ -1542,17 +1542,33 @@ TEST_F(ClusterInfoImplTest, ExtensionProtocolOptionsForUnknownFilter) {
...
+  }
+
+  std::function<ProtobufTypes::MessagePtr()> empty_proto_;
+  std::function<Upstream::ProtocolOptionsConfigConstSharedPtr(const Protobuf::Message&)> config_;
+};
+
+class TestNetworkFilterConfigFactory
+    : public Server::Configuration::NamedNetworkFilterConfigFactory {
+public:
+  TestNetworkFilterConfigFactory(TestFilterConfigFactoryBase& parent) : parent_(parent) {}
```
Comment (zuercher) [inline test/common/upstream/upstream_impl_test.cc]: It's a little odd not to use inheritance here, but I guess the alternative is assigning the lambdas to variables in the tests and re-using them.

It's a bit dodgy, but I think you could combine the two factories into a single class as long as you overload all the methods that appear in both. 😬

File: source/common/upstream/upstream_impl.cc [hunk 3]
```diff
@@ -119,12 +119,23 @@ parseExtensionProtocolOptions(const envoy::api::v2::Cluster& config) {
...
+    factory = Registry::FactoryRegistry<
+        Server::Configuration::NamedNetworkFilterConfigFactory>::getFactory(name);
+    if (factory == nullptr) {
+      factory = Registry::FactoryRegistry<
+          Server::Configuration::NamedHttpFilterConfigFactory>::getFactory(name);
+    }
+
+    if (factory == nullptr) {
+      throw EnvoyException(fmt::format(
+          "Didn't find a registered network or http implementation for name: '{}'", name));
```
Comment (zuercher) [inline source/common/upstream/upstream_impl.cc]: Let's add "filter" between http and implementation.

File: include/envoy/server/filter_config.h (hunk 1 above)
Comment (yuval-k) [reply include/envoy/server/filter_config.h]: sounds good

File: test/common/upstream/upstream_impl_test.cc (hunk 2 above)
Comment (yuval-k) [reply test/common/upstream/upstream_impl_test.cc]: The catch is that i'm not sure how that will work with `name()` method that's shared on both, but doesn't come from a common interface. is there a way to express that in c++ ?

File: source/common/upstream/upstream_impl.cc (hunk 3 above)
Comment (yuval-k) [reply source/common/upstream/upstream_impl.cc]: sounds good

File: test/common/upstream/upstream_impl_test.cc (hunk 2 above)
Comment (zuercher) [reply test/common/upstream/upstream_impl_test.cc]: It's probably a bad idea. I'm fine for you to leave this as is.

File: include/envoy/server/filter_config.h [hunk 4]
```diff
@@ -181,11 +181,35 @@ class NamedListenerFilterConfigFactory {
   virtual std::string name() PURE;
 };
 
+class ProtocolOptionsConsumer {
```
Comment (zuercher) [inline include/envoy/server/filter_config.h]: Perhaps this should be called ProtocolOptionsFactory? I'm sure what "consumer" means here.

Regardless, should add at least a brief class comment here.

**Comment by yuval-k** [comment]
@zuercher fixed 2 of the 3 comments, not sure if there's a good was to fix the suggestion on the the filter test class - see my response.

File: include/envoy/server/filter_config.h (hunk 4 above)
Comment (yuval-k) [reply include/envoy/server/filter_config.h]: Sounds good to me - will fix

Comment (yuval-k) [reply include/envoy/server/filter_config.h]: what about this comment for class doc:
`Implemented by filter factories that require more options to process the protocol used by the upstream cluster.`

**Comment by zuercher** [comment]
BTW I'm going to cancel your running builds for an older commit -- circle ci is running a backlog and I want to get master fix testing asap -- it shouldn't affect your builds for your latest (or any future) commits. Just FYI, in case you get an email.

**Comment by zuercher** [comment]
The hds integration test is being worked on in #4164.
Depending on when you branched, merging master may with the release build test (if you don't already have 07bc27c0579). You can create and push an empty commit to force another go (`git commit --allow-empty -m 'bump ci' -s`)

**Comment by zuercher** [comment]
The hds integration test is being worked on in #4164.
Depending on when you branched, merging master may with the release build test (if you don't already have 07bc27c0579). You can create and push an empty commit to force another go (`git commit --allow-empty -m 'bump ci' -s`)

**Comment by yuval-k** [comment]
The release failure also seems unrelated and stems from rbac:integration_test:

```
test/extensions/filters/network/rbac/integration_test.cc:92: Failure
Expected equality of these values:
  1U
    Which is: 1
  test_server_->counter("tcp.rbac.allowed")->value()
    Which is: 0
test/extensions/filters/network/rbac/integration_test.cc:95: Failure
Expected equality of these values:
  1U
    Which is: 1
  test_server_->counter("tcp.rbac.shadow_denied")->value()
    Which is: 0
[  FAILED  ] IpVersions/RoleBasedAccessControlNetworkFilterIntegrationTest.Allowed/IPv4, where GetParam() = 4-byte object <00-00 00-00> (53 ms)
```

**Comment by yuval-k** [comment]
yes it seems like that flaky RBAC you meantioned, i'll kick CI to have another go

**Comment by yuval-k** [comment]
Success! thanks @zuercher !

**Comment by htuch** [comment]
@yuval-k this seems reasonable, but I'd like to quickly double check on motivation here. Is it the case that your NATS streaming filter makes use of knowledge of the upstream picked cluster? Usually awareness of upstream cluster and host pick is more limited, to the router filter, earlier filters in the pipeline don't have this smarts. You can't action on this information until later in the request, post initial headers for example. 

Anyway, if you can provide some insight into how this is consume by your HTTP filter I think that would help review. Thanks.

**Comment by yuval-k** [comment]
@htuch yes that's true. In that  sense the NATS filter replaces the router filter (i.e. the NATs filter doesn't let the request continue down the filter chain). The idea is to support a case when some requests are routed to http upstreams, and others are converted to NATS pub.

**Comment by htuch** [comment]
@yuval-k how do the cluster defined protocol options fit into that idea? If you're holding the request, you probably don't even have a cluster assignment yet for it, so how can you benefit from the protocol options there?

**Comment by yuval-k** [comment]
The cluster assignment happens in the HCM before the request reaches the NATS filter (which is an http filter, placed before the router filter). 
If the NATS see that the request belongs to a NATS cluster, it prevents the request from continuing the chain and handles it. the protocol options is a way to express per cluster vs on the filter config.

**Comment by htuch** [comment]
Right, yeah; we have the cached route entry, which indicates cluster (but no actual upstream host pick yet). Makes sense.

File: test/common/upstream/upstream_impl_test.cc [hunk 5]
```diff
@@ -1625,14 +1685,27 @@ TEST_F(ClusterInfoImplTest, ExtensionProtocolOptionsForFilterWithOptions) {
       envoy.test.filter: { option: "value" }
   )EOF";
 
-  auto cluster = makeCluster(yaml);
+  std::vector<std::unique_ptr<StrictDnsClusterImpl>> clusters;
```
Comment (htuch) [inline test/common/upstream/upstream_impl_test.cc]: Can you add some comment explaining what you're testing here? It's unclear without reading deeply whether we're covering all the permutations of {no filter, network filter, http filter} for a given protocol options slot.

**Comment by yuval-k** [comment]
Not sure what happened with the DCO - I merged master to get the RBAC flake fix (did a regular `git merge origin/master`)

**Comment by htuch** [comment]
That's weird, this branch looks kind of crazy. Do you want to just get it into a sane state and force push?

**Comment by yuval-k** [comment]
sounds good, will do

**Comment by zuercher** [comment]
@yuval-k make sure you get the clang-6 change as well

**Comment by yuval-k** [comment]
I used latest master from today and cherry picked my commits over it - so it should be good.

**Pull request merged**